h2o-mapreduce-generic/src/main/java/water/hadoop/h2odriver.java

```diff
@@ -498,25 +491,188 @@ else if (type == MapperToDriverMessage.TYPE_EXIT) {
+    void announceCloudReadyNode(String leaderWebServerIp, int leaderWebServerPort) throws Exception {
+      // Do this under a synchronized block to avoid getting multiple cluster ready notification files.
+      synchronized (h2odriver.class) {
+        if (! clusterIsUp) {
+          int n = _numNodesReportingFullCloudSize.incrementAndGet();
+          if (n == _targetCloudSize) {
+            reportClusterReady(leaderWebServerIp, leaderWebServerPort);
+            clusterIsUp = true;
+          }
+        }
+      }
+    }
+
+    void announceFailedNode(String ip, int port, String hostAddress, int exitStatus) {
+      System.out.println("H2O node " + ip + ":" + port + (hostAddress != null ? " on host " + hostAddress : "") +
+              " exited with status " + exitStatus);
+      if (! clusterIsUp) {
+        clusterFailedToComeUp = true;
+      }
+    }
+
+    boolean cloudingInProgress() {
+      return !(clusterIsUp || clusterFailedToComeUp || getShutdownRequested());
+    }
+    
+    void setFlatfile(String flatfile) {
+      flatfileContent = flatfile;
+    }
+
+    final int targetCloudSize() {
+      return _targetCloudSize;
+    }
+
+    protected void fatalError(String message) {
+      System.out.println("ERROR: " + message);
+      System.exit(1);
+    }
+
+    abstract void close() throws Exception;
+
+    abstract void setMapperParameters(Configuration conf);
+  }
+
+  class FileSystemEventManager extends CloudingManager {
+
+    private volatile FSCloudingEventSource _event_source;
+
+    FileSystemEventManager(int targetCloudSize, Configuration conf, String cloudingDir) throws IOException {
+      super(targetCloudSize);
+      _event_source = new FSCloudingEventSource(conf, cloudingDir);
+    }
+
+    @Override
+    public synchronized void start() {
+      Path cloudingPath = _event_source.getPath();
+      System.out.println("Using filesystem directory to exchange information about started H2O nodes during the clouding process: " + cloudingPath);
+      System.out.println("(You can override this by adding -clouding_dir argument.)");
+      try {
+        if (! _event_source.isEmpty()) {
+          fatalError("Clouding directory `" + cloudingPath + "` already exists/is not empty.");
+        }
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      super.start();
+    }
+
+    @Override
+    void close() {
+      _event_source = null;
+    }
+
+    boolean isOpen() {
+      return _event_source != null;
+    }
+    
+    @Override
+    void setMapperParameters(Configuration conf) {
+      conf.set(h2omapper.H2O_CLOUDING_IMPL, FileSystemBasedClouding.class.getName());
+      conf.set(h2omapper.H2O_CLOUDING_DIR_KEY, _event_source.getPath().toString());
+      conf.setInt(h2omapper.H2O_CLOUD_SIZE_KEY, targetCloudSize());
+    }
+
+    private void sleep() {
+      try {
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        maybeLogException(e);
```

Comment:
same as above, interrupt should be handled and then re-interrupted

Reply:
fixed